chain/indexer/src/snapshot/download.rs

```diff
@@ -0,0 +1,434 @@
+    pub fn with_config(config: DownloadConfig) -> SnapshotResult<Self> {
+        Ok(Self {
+            client: Client::builder()
+                .timeout(config.timeout)
+                .build()
+                .map_err(SnapshotError::Network)?,
+            config,
+        })
+    }
+
+    /// Downloads a snapshot from the given URL to the target path.
+    ///
+    /// Supports HTTP/HTTPS URLs with retry logic and file:// URLs for local files.
+    ///
+    /// # Arguments
+    ///
+    /// * `url` - Source URL (http://, https://, or file:// scheme)
+    /// * `target_path` - Destination file path
+    ///
+    /// # Errors
+    ///
+    /// Returns [`SnapshotError`] for network failures, file system errors, or validation failures.
+    pub async fn download_snapshot(&self, url: &str, target_path: &Path) -> SnapshotResult<()> {
+        self.download_snapshot_with_retry(url, target_path, self.config.max_retries)
+            .await
+    }
+
+    /// Downloads a snapshot with configurable retry logic.
+    ///
+    /// Implements exponential backoff between retry attempts for HTTP/HTTPS URLs.
+    /// Local file:// URLs are handled without retry logic. Certain errors
+    /// (like 4xx HTTP status codes or insufficient disk space) will not be retried.
+    ///
+    /// # Arguments
+    ///
+    /// * `url` - The HTTP/HTTPS or file:// URL to download/copy from
+    /// * `target_path` - Local path where the downloaded file will be saved
+    /// * `max_retries` - Maximum number of retry attempts (ignored for file:// URLs)
+    ///
+    /// # Errors
+    ///
+    /// Returns `SnapshotError` for various failure conditions including:
+    /// - Network errors (with retry)
+    /// - HTTP errors (4xx without retry, 5xx with retry)
+    /// - Insufficient disk space (without retry)
+    /// - File size exceeding limits (without retry)
+    pub async fn download_snapshot_with_retry(
+        &self,
+        url: &str,
+        target_path: &Path,
+        max_retries: u32,
+    ) -> SnapshotResult<()> {
+        let backoff = backon::ExponentialBuilder::default().with_max_times(max_retries as usize);
+
+        struct Sleeper;
+        impl backon::Sleeper for Sleeper {
+            type Sleep = futures_timer::Delay;
+
+            fn sleep(&self, dur: Duration) -> Self::Sleep {
+                futures_timer::Delay::new(dur)
+            }
+        }
+
+        (|| async { self.download_snapshot_once(url, target_path).await })
+            .retry(backoff)
+            .sleep(Sleeper)
+            .when(|err| {
+                !matches!(
+                    err,
+                    SnapshotError::TooLarge { .. }
+                        | SnapshotError::InsufficientSpace { .. }
+                        | SnapshotError::HttpStatus { status: 400..=499 },
+                )
+            })
+            .notify(|err, _dur| {
+                warn!("Download attempt failed: {}", err);
```

Comment:
```suggestion
                warn!(error = %err, "Download attempt failed");
```

Reply:
Even better: name the variable `error` and then just do `warn!(%error, "Downaload attempt failed")`

Btw are you sure this should be a warning and not an error rather?

Reply:
Fixed.